src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,190 @@
+        _nest_path: list[str],
+        _type: Callable[[str], _T] | None = None,
+        choices: Iterable[_T] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: The name of the ConfigOption
+            env_name: Environmental variable value should be read from, if not supplied, we'll construct this
+            type: A function that can turn str to the desired type, useful for reading value from environmental variable
+        """
+        self.name = name
+        self.type = _type
+        self.choices = choices
+        self._nest_path = _nest_path + [name]
+        self._root_parser: ConfigParser = _root_parser
+        self.env_name = env_name
+
+    def get(self):
+        """Retrieve a value of option."""
+        source = "environment variable"
+        loaded_env, value = self._get_env()
+        if not loaded_env:
+            source = "configuration file"
+            value = self._get_config()
+        if self.choices and value not in self.choices:
+            raise ConfigSourceError(
+                f"The value of {self.option_name} read from "
+                f"{source} is not part of {self.choices}"
+            )
+        return value
+
+    @property
+    def option_name(self) -> str:
+        return ".".join(self._nest_path[1:])
+
+    @property
+    def option_env_name(self) -> str:
+        pieces = map(lambda e: e.upper(), self._nest_path[1:])
+        return f"SF{'_' + '_'.join(pieces)}"
+
+    def _get_env(self) -> tuple[bool, str | _T | None]:
+        if self.env_name is False:
+            return False, None
+        if self.env_name is not None:
+            env_name = self.env_name
+        else:
+            # Generate environment name if it wasn't not explicitly supplied,
+            #  and isn't disabled
+            env_name = self.option_env_name
+        if env_name not in os.environ:
+            return False, None
+        env_var = os.environ.get(env_name, None)
+        loaded_var: _T | None | str = env_var
+        if env_var and self.type is not None:
+            loaded_var = self.type(env_var)
+        if isinstance(loaded_var, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return True, loaded_var.value
+        return True, loaded_var
+
+    def _get_config(self):
+        e = self._root_parser._conf
+        for k in self._nest_path[1:]:
+            e = e[k]
+        if isinstance(e, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return e.value
+        return e
+
+
+class ConfigParser:
+    def __init__(
+        self,
+        *,
+        name: str,
+        file_path: Path | None = None,
+    ):
+        self.name = name
+        self.file_path = file_path
+        # Objects holding subparsers and options
```

Comment:
Could you explain what's subparser? parser for different sections in a file?

Reply:
Yes exactly. One example:
```toml
[connections.preprod]
user = "user"
account = "acccount"
password = "password"

[snowcli.output]
format = "json"
verbose = true
```

Refers to:
```
Root parser
|
snowcli
|
output
|               \
(format)   (verbose)
```
So being able to nest `ConfigParser`s in each other allows for this

Reply:
I didn't realize you needed to do this, or is it just a convenience?  I wonder if using a true [TOML parser](https://docs.python.org/3/library/tomllib.html) would make things simpler?

Reply:
This isn't because of TOML, this paradigm arose from the fact that I want to nest things not only in TOML, but also in environmental variables and possibly even more.
I use this hierachy to automatically generate environmental variables like: `SNOWFLAKE_SNOWCLI_OUTPUT_VERBOSE` from my example and I forsee using this very same thing to eventually parse arguments from SnowSQL(or other CLI) CLI arguments. So my example could be `snowsql --output_format json`.

Reply:
I wish we had https://github.com/spf13/viper for python

Reply:
I've never used it, but something like https://github.com/alexferl/vyper ?